language/stdlib/modules/DualAttestation.move

```diff
@@ -0,0 +1,308 @@
+        pragma opaque = true, verify = false;
+        ensures result == spec_dual_attestation_message(payer, metadata, deposit_value);
+    }
+    spec module {
+        /// Uninterpreted function for `Self::dual_attestation_message`.
+        define spec_dual_attestation_message(payer: address, metadata: vector<u8>, deposit_value: u64): vector<u8>;
+    }
+
+    /// Helper function to check validity of a signature when dual attestion is required.
+    public fun assert_signature_is_valid(
+        payer: address,
+        payee: address,
+        metadata_signature: vector<u8>,
+        metadata: vector<u8>,
+        deposit_value: u64
+    ) acquires Credential {
+        // sanity check of signature validity
+        assert(Vector::length(&metadata_signature) == 64, EMALFORMED_METADATA_SIGNATURE);
+        // cryptographic check of signature validity
+        let message = dual_attestation_message(payer, metadata, deposit_value);
+        let credential_address = if (VASP::is_child(payee)) VASP::parent_address(payee) else payee;
+        assert(
+            Signature::ed25519_verify(
+                metadata_signature,
+                compliance_public_key(credential_address),
+                message
+            ),
+            EINVALID_METADATA_SIGNATURE
+        );
+    }
+    spec fun assert_signature_is_valid {
+        pragma verify = true, opaque = true;
+        aborts_if !exists<Credential>(payee);
+        aborts_if !signature_is_valid(payer, payee, metadata_signature, metadata, deposit_value);
+    }
+    spec module {
+        /// Returns true if signature is valid.
+        define signature_is_valid(
+            payer: address,
+            payee: address,
+            metadata_signature: vector<u8>,
+            metadata: vector<u8>,
+            deposit_value: u64
+        ): bool {
+            len(metadata_signature) == 64
+                && Signature::spec_ed25519_verify(
+                        metadata_signature,
+                        spec_compliance_public_key(payee),
+                        spec_dual_attestation_message(payer, metadata, deposit_value)
+                   )
+        }
+    }
+
+    /// Public API for checking whether a payment of `value_microlibra` coins of type `Currency`
+    /// from `payer` to `payee` has a valid dual attestation. This returns without aborting if
+    /// (1) dual attestation is not required for this payment, or
+    /// (2) dual attestation is required, and `metadata_signature` can be verified on the message
+    ///     `metadata` | `payer` | `amount` | `DOMAIN_SEPARATOR` using the `compliance_public_key`
+    ///     published in `payee`'s `Credential` resource
+    /// It aborts with an appropriate error code if dual attestation is required, but one or more of
+    /// the conditions in (2) is not met.
+    public fun assert_payment_ok<Currency>(
+        payer: address,
+        payee: address,
+        value_microlibra: u64,
```

Comment:
Maybe `value` instead of `value_microlibra`? If I understand correctly this value is in `Currency`.